pennylane/hf/tapering.py

```diff
@@ -0,0 +1,204 @@
+
+    Args:
+        terms (Iterable[Observable]): operators defining the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+    Returns:
+        E (ndarray): binary matrix representation of the Hamiltonian of shape
+        :math:`len(terms) \times 2*num_qubits`.
+
+    .. code-block::
+        >>> terms = [PauliZ(wires=[0]) @ PauliX(wires=[1]), PauliZ(wires=[0]) @ PauliY(wires=[2]),
+                        PauliXwires=[0]) @ PauliY(wires=[3])]
+        >>> get_binary_matrix(terms, 4)
+         array([[1, 0, 0, 0, 0, 1, 0, 0],
+                [1, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 0, 1, 1, 0, 0, 1]]))
+
+    """
+
+    E = np.zeros((len(terms), 2 * num_qubits), dtype=int)
+    for idx, term in enumerate(terms):
+        ops, wires = term.name, term.wires
+        if len(term.wires) == 1:
+            ops = [ops]
+        for op, wire in zip(ops, wires):
+            if op in ["PauliX", "PauliY"]:
+                E[idx][wire + num_qubits] = 1
+            if op in ["PauliZ", "PauliY"]:
+                E[idx][wire] = 1
+
+    return E
+
+
+def _reduced_row_echelon(binary_matrix):
+    """ Returns the reduced row echelon form (RREF) of a matrix in a binary finite field Z_2 """
+
+    binary_matrix = deepcopy(binary_matrix)
+    shape = binary_matrix.shape
+
+    for irow, icol in zip(range(shape[0]), range(shape[1])):
+
+        # find value and index of largest element in remainder of column icol
+        krow = irow + np.argmax(binary_matrix[irow:, icol])
+
+        # swap rows krow and irow
+        binary_matrix[krow], binary_matrix[irow] = deepcopy(binary_matrix[irow]), deepcopy(binary_matrix[krow])
+
+        # store remainder columns of the row irow
+        pvtcols = binary_matrix[irow, icol:]
+
+        # get the column icol and set its irow element to 0 to avoid XORing pivot row with itself
+        currcol = deepcopy(binary_matrix[:, icol])
+        currcol[irow] = 0
+        binary_matrix[:, icol:] ^= np.outer(currcol, pvtcols)
+
+    return binary_matrix.astype(int)
+
+
+def _kernel(binary_matrix):
+    """ Computes the kernel of a binary matrix on the binary finite field Z_2 """
+
+    # Get the columns with and without pivots
+    pivots = (binary_matrix.T!=0).argmax(axis=0)
+    nonpivots = np.setdiff1d(range(len(binary_matrix[0])), pivots)
+
+    # Initialize the nullspace
+    null_vector = np.zeros((binary_matrix.shape[1], len(nonpivots)), dtype=int)
+    null_vector[nonpivots, np.arange(len(nonpivots))] = 1
+
+    # Fill up the nullspace vectors from the binary matrix
+    null_vector_indices = np.ix_(pivots, np.arange(len(nonpivots)))
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices]%2
+
+    return null_vector.T
+
+
+def generate_taus(nullspace, num_qubits):
+    """Generate generators tau from the nullspace
```

Comment:
```suggestion
    """Compute the tau generators from the nullspace of the binary matrix form of a Hamiltonian
```
I wonder if it's worth briefly explaining what the tau generators are. Is there a paper where they are defined?

Reply:
I have tried to give a brief definition and a link to the paper as well.